pkgs/build-support/flatten-references-graph/src/flatten_references_graph/split_paths.py

```diff
@@ -0,0 +1,226 @@
+from toolz import curried as tlz
+from toolz import curry
+
+from .lib import (
+    debug,
+    debug_plot,
+    DEBUG_PLOT,
+    find_vertex_by_name_or_none,
+    graph_is_empty,
+    is_None,
+    subcomponent_multi,
+    unnest_iterable
+)
+
+
+@curry
+def add_root(added_root_name, graph):
+    """Add single root to the graph connected to all existing roots.
+
+    If graph has only one root, return the graph unchanged and the name
+    of the root vertex.
+
+    Otherwise return a modified graph (copy) and a name of the added root
+    vertex.
+    """
+    roots = graph.vs.select(lambda v: len(graph.predecessors(v)) == 0)
+    root_names = roots["name"]
+
+    if len(root_names) == 1:
+        return graph, root_names[0]
+    else:
+        edges = [(added_root_name, v) for v in root_names]
+        graph_with_root = graph + added_root_name + edges
+        return graph_with_root, added_root_name
+
+
+@curry
+def remove_added_root(added_root_name, graph):
+    """Remove vertex with given name, returning copy of input graph if vertex
+    with given name is found in the graph
+    """
+    added_root = find_vertex_by_name_or_none(graph)(added_root_name)
+
+    return graph - added_root_name if added_root else graph
+
+
+def get_children_of(graph, vertex_names):
+    return unnest_iterable(map(
+        graph.successors,
+        tlz.remove(
+            is_None,
+            map(
+                find_vertex_by_name_or_none(graph),
+                vertex_names
+            )
+        )
+    ))
+
+
+def as_list(x):
+    return x if isinstance(x, list) else [x]
+
+
+@curry
+def split_path_spec_to_indices(graph, split_path_spec):
+    debug("split_path_spec", split_path_spec)
+    if isinstance(split_path_spec, dict):
+        if "children_of" in split_path_spec:
+            children_of = split_path_spec["children_of"]
+
+            return get_children_of(graph, as_list(children_of))
+        else:
+            raise Exception(
+                "Unexpected split path spec: dict with invalid keys."
+                "Valid: [\"children_of\"]"
+            )
+    else:
+        vertex = find_vertex_by_name_or_none(graph)(split_path_spec)
+        return [] if is_None(vertex) else [vertex.index]
+
+
+call_count = 0
+
+
+@curry
+def split_paths(split_paths, graph_in):
+    debug("____")
+    debug("split_paths:", split_paths)
+    debug("graph_in:", graph_in)
+
+    if DEBUG_PLOT:
+        global call_count
+        graph_name_prefix = f"split_paths_{call_count}_"
+        call_count += 1
```

Comment:
As far as I can see these `call_count`'s are unused? Why are we keeping track of them only when `DEBUG_PLOT` is True?

Reply:
`call_count` is use, on line 93 just above

`graph_name_prefix` is then used in multiple times below (only if `DEBUG_PLOT` is enabled )

we need unique name for each call since we are [generating and saving images of the grapshs](https://github.com/NixOS/nixpkgs/pull/122608/files#diff-6bf554b00f6a3f75421b7b7cfa75a0b22a991f553856f4e965a8bf5d541d538eR38-R45) and we want new image for each call